### PR TITLE
Structure_Common_oM: Updates to IResultCollection and MeshResult

### LINKED ISSUE #589: Structure_oM: Update MeshResult and IResultCollection to be IResult
#### Description:

After discussions [here](https://github.com/BHoM/BHoM/pull/587#issuecomment-544912049) about general Result handling and adapter ease of use, some changes are proposed to the IResultCollection and in turn the MeshResults.

Proposed changes are:

1. Make `IResultCollection` implement `IResult`
1. Have one empty `IResultCollection` interface and a Generic one. First for filter, second for generic helper methods. Proposed code:

```
namespace BH.oM.Common
{

    public interface IResultCollection : IObject, IImmutable, IResult
    { 

    }

    public interface IResultCollection<T> : IResultCollection where T : IResult
    {
        /***************************************************/
        /**** Properties                                ****/
        /***************************************************/

        ReadOnlyCollection<T> Results { get; }

        /***************************************************/
    }
}
```
3. Update names of `MeshResults` and `MeshResult` classes to avoid confusion. Proposal to rename `MeshResult` to `MeshElementResult` and `MeshResults` to `MeshResult.
4. The new `MeshResult` will need additional attributes given that it is now implementing `IResult` itself. Proposed code:

```
public class MeshResult : IResult, IImmutable, IResultCollection<MeshElementResult>
    {
        /***************************************************/
        /**** Properties                                ****/
        /***************************************************/

        public IComparable ObjectId { get; } = "";

        public IComparable ResultCase { get; }

        public double TimeStep { get; }

        public MeshResultLayer Layer { get;} 

        [Description("Position within the element thickness that result is extracted from, normalised to 1. 0 = lower surface, 0.5 = middle, 1 = top surface")]
        public double LayerPosition { get; }

        public MeshResultSmoothingType Smoothing { get;  }

        public ReadOnlyCollection<MeshElementResult> Results { get; }


        /***************************************************/
        /**** Constructors                              ****/
        /***************************************************/

        public MeshResult(IComparable objectId, IComparable resultCase, double timeStep, MeshResultLayer resultLayer, MeshResultSmoothingType smoothing, ReadOnlyCollection<MeshElementResult> results)
        {
            ObjectId = objectId;
            ResultCase = resultCase;
            TimeStep = timeStep;
            Layer = resultLayer;
            LayerPosition = LayerPosition;
            Smoothing = smoothing;
            Results = results;
        }
```

This will simplify adapter methods as well as queries.

@al-fisher @rwemay @alelom

### PR DESCRIPTION
**Pull request by IsakNaslundBh**
### Issues addressed by this PR

Closes #589 

### Changelog

- Making `IResultCollection : IResult`
- Renaming MeshReslt to MeshElementResult
- Renaming MeshResults to MeshResult
- Implementing IResult interface on new MeshResult and updating property names

### Additional comments

This will break structural structural adapters capable of extracting mesh results. All these need to be updated before merging.

Only Etabs and Robot toolkit had any mesh results implemented and have now been updated. This should be good to go. as soon as all PRs are approved

### CODE REVIEW

File: Common_oM/Results/IResultCollection.cs [hunk 1]
```diff
@@ -22,17 +22,28 @@
...
+    /***************************************************/
+
+    public interface IResultCollection : IObject, IImmutable, IResult
+    {
+
+    }
+
+    /***************************************************/
+
+    public interface IResultCollection<T> : IResultCollection where T : IResult
```
Comment (alelom) [inline Common_oM/Results/IResultCollection.cs]: 👍

**Comment by FraserGreenroyd** [comment]
@IsakNaslundBh while I note the changes are to both `Common` and `Structure`, I wonder if the PR title should start with `Structure` so that it gets grouped with the other `Structure` items in the change log when it's produced given that the change affects `Structure` users mostly?

Just a thought!

**Comment by IsakNaslundBh** [comment]
> @IsakNaslundBh while I note the changes are to both Common and Structure, I wonder if the PR title should start with Structure so that it gets grouped with the other Structure items in the change log when it's produced given that the change affects Structure users mostly?

Thanks @FraserGreenroyd . Think you are right. My first reasoning was in the opposite direction, put common first as it has the possibility to affect more than just structures, but as only Structures is affected so far, I have changed to what you suggested:)

**Comment by al-fisher** [comment]
Yep - looks good to me too - merging with 
https://github.com/BHoM/BHoM_Engine/pull/1322 
https://github.com/BHoM/Robot_Toolkit/pull/279 
https://github.com/BHoM/ETABS_Toolkit/pull/228

**Pull request merged**